sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/ContainerRegistryClientBuilder.java

```diff
@@ -0,0 +1,290 @@
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+/** A builder for creating a new instance of the ContainerRegistry type. */
+@ServiceClientBuilder(
+        serviceClients = {
+            ContainerRegistryClient.class,
+            ContainerRegistryAsyncClient.class
+        })
+public final class ContainerRegistryClientBuilder {
+    private static final String SDK_NAME = "name";
+    private static final String SDK_VERSION = "version";
+
+    private static final Map<String, String> PROPERTIES =
+        CoreUtils.getProperties("azure-containers-containerregistry.properties");
+
+    private final ClientLogger logger = new ClientLogger(ContainerRegistryClientBuilder.class);
+    private final List<HttpPipelinePolicy> perCallPolicies = new ArrayList<>();
+    private final List<HttpPipelinePolicy> perRetryPolicies = new ArrayList<>();
+    private ClientOptions clientOptions;
+    private Configuration configuration;
+    private String endpoint;
+    private HttpClient httpClient;
+    private TokenCredential credential;
+    private HttpPipeline httpPipeline;
+    private HttpLogOptions httpLogOptions;
+    private SerializerAdapter serializerAdapter;
+    private RetryPolicy retryPolicy;
+
+    /**
+     * Sets Registry login URL.
+     *
+     * @param endpoint the endpoint value.
+     * @return the ContainerRegistryBuilder.
+     */
+    public ContainerRegistryClientBuilder endpoint(String endpoint) {
+        try {
+            new URL(endpoint);
+        } catch (MalformedURLException ex) {
+            throw logger.logExceptionAsWarning(new IllegalArgumentException("'endpoint' must be a valid URL"));
+        }
+
+        this.endpoint = endpoint;
+        return this;
+    }
+
+    /**
+     * Sets Registry login URL.
+     *
+     * @param credential the credential to use to access registry.
+     * @return the ContainerRegistryBuilder.
+     */
+    public ContainerRegistryClientBuilder credential(TokenCredential credential) {
+        this.credential = Objects.requireNonNull(credential,
+            "'credential' cannot be null.");
+        return this;
+    }
+
+    /**
+     * Sets The HTTP httpPipeline to send requests through.
+     *
+     * @param httpPipeline the httpPipeline value.
+     * @return the ContainerRegistryBuilder.
+     */
+    public ContainerRegistryClientBuilder httpPipeline(HttpPipeline httpPipeline) {
+        if (this.httpPipeline != null && httpPipeline == null) {
+            logger.info("HttpPipeline is being set to 'null' when it was previously configured.");
+        }
+
+        this.httpPipeline = httpPipeline;
+        return this;
+    }
+
+    /**
+     * Sets The serializer to serialize an object into a string.
+     *
+     * @param serializerAdapter the serializerAdapter value.
+     * @return the ContainerRegistryBuilder.
+     */
+    public ContainerRegistryClientBuilder serializerAdapter(SerializerAdapter serializerAdapter) {
```

Comment:
Generally, we don't allow a `SerializerAdapter` to be configured in clients as generated code only works with `JacksonAdapter`. Was this added to support scenarios where the SDK needs to support serialization of customer owned types? If so, you'll want to look into `ObjectSerializer`/`JsonSerializer`, use the Azure Search library as an example of it.

Reply:
OOh.. I did not know that. Thanks! No special need from the service. I saw a couple of examples in our libraries and took it from there.

---
In reply to: [602632006](https://github.com/Azure/azure-sdk-for-java/pull/20139#discussion_r602632006) [](ancestors = 602632006)